document/src/vespa/document/update/documentupdate.cpp

```diff
@@ -165,120 +206,179 @@ DocumentUpdate::serializeFlags(int size_) const
+    // We need type for object to work, so just throwing exception if it's
+    // not there.
+    if((content & CONTENT_HASTYPE) == 0) {
+        throw IllegalStateException("Missing document type", VESPA_STRLOC);
+    }
 
-        vespalib::stringref typestr = buffer.getBufferAtPos();
-        buffer.incPos(typestr.length() + 1);
+    vespalib::stringref typestr = stream.peek();
+    stream.adjustReadPos(typestr.length() + 1);
 
-        int16_t version = 0;
-        buffer.getShortNetwork(version);
-        const DocumentType *type = repo.getDocumentType(typestr);
-        if (!type) {
-            throw DocumentTypeNotFoundException(typestr, VESPA_STRLOC);
-        }
-        return std::make_pair(type, docId);
+    int16_t version = 0;
+    stream >> version;
+    const DocumentType *type = repo.getDocumentType(typestr);
+    if (!type) {
+        throw DocumentTypeNotFoundException(typestr, VESPA_STRLOC);
     }
+    return std::make_pair(type, docId);
+}
 }
 
 // Deserialize the content of the given buffer into this document update.
 DocumentUpdate::UP
-DocumentUpdate::create42(const DocumentTypeRepo& repo, ByteBuffer& buffer)
+DocumentUpdate::create42(const DocumentTypeRepo& repo, vespalib::nbostream & stream)
 {
-    return std::make_unique<DocumentUpdate>(repo, buffer, SerializeVersion::SERIALIZE_42);
+    auto update = std::make_unique<DocumentUpdate>();
+    update->init42(repo, stream);
+    return update;
 }
 
 DocumentUpdate::UP
 DocumentUpdate::createHEAD(const DocumentTypeRepo& repo, ByteBuffer& buffer)
 {
-    return std::make_unique<DocumentUpdate>(repo, buffer, SerializeVersion::SERIALIZE_HEAD);
+    vespalib::nbostream is(buffer.getBufferAtPos(), buffer.getRemaining());
+    auto update = std::make_unique<DocumentUpdate>();
+    update->initHEAD(repo, is);
+    buffer.setPos(buffer.getPos() + is.rp());
+    return update;
+}
+
+DocumentUpdate::UP
+DocumentUpdate::createHEAD(const DocumentTypeRepo& repo, vespalib::nbostream stream)
+{
+    auto update = std::make_unique<DocumentUpdate>();
+    update->initHEAD(repo, std::move(stream));
+    return update;
 }
 
 void
-DocumentUpdate::deserialize42(const DocumentTypeRepo& repo, ByteBuffer& buffer)
+DocumentUpdate::init42(const DocumentTypeRepo & repo, vespalib::nbostream & stream)
 {
-    int pos = buffer.getPos();
-    try{
-        buffer.getShortNetwork(_version);
+    _repo = &repo;
+    deserialize42(repo, stream);
+    reserialize();
+}
+void
+DocumentUpdate::initHEAD(const DocumentTypeRepo & repo, vespalib::nbostream && stream)
+{
+    _repo = &repo;
+    _backing = std::move(stream);
+    size_t startPos = _backing.rp();
+    deserializeHeader(repo, _backing);
+    _backing.rp(startPos);
+}
+
+void
+DocumentUpdate::initHEAD(const DocumentTypeRepo & repo, vespalib::nbostream & stream)
+{
+    size_t startPos = stream.rp();
+    deserializeHEAD(repo, stream);
+    size_t sz = stream.rp() - startPos;
+    _backing = nbostream(stream.peek() - sz, sz);
```

Comment:
No, this is intentional. If you hand over the stream the state of stream is of no importance to the caller.
However there is still one place where it can not be lazy. When arriving over documentapi no length information is available so the caller can continue and deserialize the stuff after the update unless the stream for the whole message is consumed serially.

It would be nice to have this fixed so we can use handover and have it always lazy, but I did not immediately see any solution.